federatedscope/core/data/base_translator.py

```diff
@@ -0,0 +1,129 @@
+import logging
+import numpy as np
+
+from federatedscope.core.auxiliaries.splitter_builder import get_splitter
+from federatedscope.core.data import ClientData, StandaloneDataDict
+
+logger = logging.getLogger(__name__)
+
+
+class BaseDataTranslator:
+    """
+    Perform process:
+        Dataset -> ML split -> FL split -> Data (passed to FedRunner)
+
+    """
+    def __init__(self, global_cfg, client_cfgs=None):
+        """
+        Convert data to `StandaloneDataDict`.
+
+        Args:
+            global_cfg: global CfgNode
+            client_cfgs: client cfg `Dict`
+        """
+        self.global_cfg = global_cfg
+        self.client_cfgs = client_cfgs
+        self.splitter = get_splitter(global_cfg)
+
+    def __call__(self, dataset):
+        """
+
+        Args:
+            dataset: `torch.utils.data.Dataset`, `List` of (feature, label)
+                or split dataset tuple of (train, val, test) or Tuple of
+                split dataset with [train, val, test]
+
+        Returns:
+            datadict: instance of `StandaloneDataDict`, which is a subclass of
+            `dict`.
+        """
+        datadict = self.split(dataset)
+        datadict = StandaloneDataDict(datadict, self.global_cfg)
+
+        return datadict
+
+    def split(self, dataset):
+        """
+        Perform ML split and FL split.
+
+        Returns:
+            dict of `ClientData` with client_idx as key.
+
+        """
+        train, val, test = self.split_train_val_test(dataset)
+        datadict = self.split_to_client(train, val, test)
+        return datadict
+
+    def split_train_val_test(self, dataset):
```

Comment:
So what is the recommended way to implement customized split funcs? e.g., VMF and HMF datasets need to be split according to user/item ids.

Reply:
In that way, we use a `dummysplitter`, where we treat MF datasets as FL datasets.